### DISCUSSION TITLE
New error since 1.1.347: Cannot pass a union of types instead of supertype

### DISCUSSION THREAD

**Discussion by superlopuh** [microsoft/pyright#7000] [Q&A] [answered]
I'm not sure if I'm using generics incorrectly, and Pyright has highlighted a bug in my code, so I thought I'd ask a question in the discussion rather than file an issue. Here's a minimal reproducer:

``` python
class Super:
    ...

class A(Super):
    ...

class B(Super):
    ...

def func(arg: type[Super]):
    ...

"""
hello.py
  hello.py:17:6 - error: Argument of type "type[A] | type[B]" cannot be assigned to parameter "arg" of type "type[Super]" in function "func"
    Type "type[UnionType]" cannot be assigned to type "type[Super]" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
"""
func(A | B)
```

In my understanding, `type[Super]` is covariant, and accepts a subtype of `Super` also. A union of subtypes of `Super` should also be a subtype of `Super`. So there shouldn't be any errors raised. Am I misunderstanding something, or is this a bug?

I had a search for some of the keywords in the title but couldn't find anything

**Comment by erictraut** [comment, marked as answer]
This is intended behavior. This change was made to tighten up pyright's behavior with regard to the official typing spec.

The problem with pyright's old behavior here is that the implementation of `func` may assume that it can instantiate `arg` by calling it.

```python
def func(arg: type[Super]):
    arg()
```

But that would produce a runtime error because the value you're passing to `arg` isn't an instantiable class. Rather, it's a runtime `UnionType` object.

It is legal to pass a value whose type is a union of two types.
```python
def func2(arg: type[A] | type[B]):
    func(arg)
```

**Reply by superlopuh** [reply]
Thank you for the explanation

**Reply by math-fehr** [reply]
@erictraut, is there a workaround like [TypeForm](https://github.com/python/mypy/issues/9773) that would allow us to have the previous semantics/meaning?

One use case we have is that we implement our own `isinstance`, with the following signature:

```python
from typing import TypeGuard, Any
T = TypeVar("T")

def my_isinstance(a: Any, t: type[T]) -> TypeGuard[T]:
    ...
```

Which allowed things like `my_isinstance(a, int | float)`.

Is there a way to have a similar signature with the more recent pyright version? While it would be possible to have overrides like `my_isinstance(a: Any, t: type[T], u: type[U]) -> TypeGuard[T | U]`, we would prefer avoiding that as we have many functions that used to take that `t: type[T]` input.

**Reply by math-fehr** [reply]
And also, something I do not understand with your example, how do we know that the `arg` passed does not have an overridden constructor? Here is a case where `pyright` does not return any error, while there is a runtime error because of a missing argument:

```python
class A:
    def __init__(self, a: int):
        print(a)


class B(A):
    def __init__(self, a: int, b: float):
        super().__init__(a)
        print(b)


def foo(a: type[A]):
    a(2)
```

**Reply by erictraut** [reply]
`TypeForm` has never been formalized or accepted into the typing spec, so that's not an option currently. If you would like to champion that as an extension to the typing system, you're welcome to do so.

With the current typing system, I don't see a way to support unions. You could support tuples of types in your custom `isinstance` through the use of overloads.

```python
from typing import TypeGuard, Any, TypeVar, overload

T1 = TypeVar("T1")
T2 = TypeVar("T2")

@overload
def my_isinstance(a: object, t: type[T1]) -> TypeGuard[T1]:
    ...

@overload
def my_isinstance(a: object, t: tuple[type[T1], type[T2]]) -> TypeGuard[T1 | T2]:
    ...

def func(a: object):
    assert my_isinstance(a, (int, str))
    reveal_type(a)  # int | str
```

> And also, something I do not understand with your example, how do we know that the arg passed does not have an overridden constructor?

That's a hole in the current typing spec. Actually, it's an implicit hole because the typing spec doesn't currently specify that `__new__` and `__init__` are exempt from LSP checks. All major type checkers currently treat it that way, and the spec should probably state that explicitly. But that means there's no enforcement of type consistency for constructor overrides. If you have a suggestion for how to improve this in the type system, please post to the typing forum.

**Reply by math-fehr** [reply]
Thanks a lot for your answer, I'll try to take a look at what is needed to make `TypeForm` happen (it seems that someone is also interested to continue this work).
